### DISCUSSION TITLE
one input to three output

### DISCUSSION THREAD

**Discussion by FFiot** [pschatzmann/arduino-audio-tools#265] [Q&A] [answered]
**Function**

**Method**
1. Is there a way not modify the kernel code？
2. Build a AudioStream class for data callback，do something,  call to->write。
3. Do something with <StreamCopyT>  
    @ lib\arduino-audio-tools\src\AudioTools\AudioCopy.h
Add code:
    StreamCopyT(Print *to_array, size_t to_num, Stream &from,  int buffer_size=DEFAULT_BUFFER_SIZE);
    Print *to_array = nullptr;   （each "to" need delayCount.）
    size_t to_num;
copy and modify: 
A) void setCallbackOnWrite(void (*onWrite)(void*obj, void*buffer, size_t len), void* obj);
    to
    void addCallbackOnWrite(void (*onWrite)(void*obj, void*buffer, size_t len), void* obj);
B) inline size_t copy(); 
    to 
    inline size_t copyMany()；
    modify L116：result = write(bytes_read, delayCount);
C) Need some infor:
    L119: if (onWrite!=nullptr) onWrite(onWriteObj, buffer, result);
    setCallbackOnWrite() only called by AudioPlay.h Line:634.

**Comment by pschatzmann** [comment, marked as answer]
I don't think this is an issue but should be rather a discussion and I suggest you read the documentation: https://github.com/pschatzmann/arduino-audio-tools/wiki/Splitting-and-Merging-Audio.  A MultiOutput with an output to ConvertedStream should be your starting point.
- You will need to implement your own Converter to get the data of the left or right channel only. Alternatively you can use the ConverterFillLeftAndRight to copy a specific channel to the other. The current ChannelReducer is mixing all channels into one.
- Currently there is no AudioStream class with a callback: However you can use the ConvertedStream class by providing your own converter.
- Though you can write your own copy class, I don't think that it make any sense to me. If the delay is missing just write your own delay converter. Since you only  process one channel you can investigate to use an Effect chain instead (https://github.com/pschatzmann/arduino-audio-tools/wiki/Audio-Effects)  and use the existing Delay effect

From my experience it is already quite a challenge to implement a simple A2DP source and make it stable. Adding all this complexity will make it even more difficult...

**Reply by FFiot** [reply]
Thank you very much. I have taken about 2 weeks on A2DP of esp32......

A2DP of ESP32 is not an excellent wireless audio solution for stethoscopes， which is too old and unreliable.
And Espressif abandons classic Bluetooth.

I will use esp32S3 with wired Headphones and wait for BLE audio.

Two problems， do not care about：

It is not working after 1 minute.
Input: i2s； Output i2s, A2DP.
Board: ESP32 Audio Kit V2.2 A149, ES8388.
Code: https://github.com/FFiot/ESP32DEMO/blob/20220717Multiple/src/main.cpp

Can not connect A2DP sink name with space like "ROD RAIN 5.0PRO".
